backends/p4tools/modules/testgen/targets/bmv2/test/testgen_api/control_plane_filter_test.cpp

```diff
@@ -0,0 +1,124 @@
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "test/gtest/helpers.h"
+
+#include "backends/p4tools/modules/testgen/options.h"
+#include "backends/p4tools/modules/testgen/targets/bmv2/test_backend/protobuf_ir.h"
+#include "backends/p4tools/modules/testgen/testgen.h"
+
+namespace Test {
+
+namespace {
+
+TEST(P4TestgenControlPlaneFilterTest, FiltersControlPlaneEntities) {
+    std::stringstream streamTest;
+    streamTest << R"p4(
```

Comment:
Maybe you make these manually every time... But would it be worth just having a function that returns something like this for all of your tests?

Regardless, you're probably better off putting this in a function here and then calling it in multiple tests. It would make it more natural to test e.g. what happens if you're given a non-existent control plane entity. Or one that isn't supported yet?

It looks like it would just be ignored effectively. Is that what you want to have happen? Do you want to do greater validation and give an Unimplemented error or an InvalidArgument error depending on the situation?

Reply:
>Maybe you make these manually every time... But would it be worth just having a function that returns something like this for all of your tests?

Yes! I thought about this, but I figured I should wait until have more clarity on the test patterns, e.g., what elements and headers I keep reusing. 

>It looks like it would just be ignored effectively. Is that what you want to have happen? 

So this was intentional. Mostly because P4Testgen currently does not have a collection of valid control-plane elements. I first would need to create that collection in a target-independent manner (since BMv2 has different control-plane elements to Tofino or the DPDK-target), then perform the lookup. Presumably I could do this with a P4Info, but again not every extension uses that either.

It would be nice to have that validation, but it will require some work.

Reply:
I see. It seems fine to skip the validation (I think that's fairly typical), but I would prefer to have some tests with those cases showing that nothing catastrophic happens at least. I think that's the only change I really would want to see.

And again, I would recommend doing it (at least locally in this file) by factoring out the P4 program creating to a function that you call in ~4 different test cases (showing that it works with 1 or multiple tables, and the behavior with a not-yet-supported and non-existent entity).